buffer/src/main/java/io/netty/buffer/api/internal/LeakDetection.java

```diff
@@ -0,0 +1,137 @@
+        try {
+            LEAK_DETECTION_ENABLED_UPDATER = MethodHandles.lookup().findStaticVarHandle(
+                    LeakDetection.class, "leakDetectionEnabled", int.class);
+        } catch (Exception e) {
+            throw new ExceptionInInitializerError(e);
+        }
+        if (enabled > 0) {
+            CALLBACKS.add(new LoggingLeakCallback());
+        }
+        leakDetectionEnabled = enabled;
+    }
+
+    private LeakDetection() {
+    }
+
+    public static AutoCloseable onLeakDetected(Consumer<LeakInfo> callback) {
+        LEAK_DETECTION_ENABLED_UPDATER.getAndAddAcquire(1);
+        Set<Consumer<LeakInfo>> callbacks = CALLBACKS;
+        synchronized (callbacks) {
+            callbacks.add(callback);
+        }
+        return new CallbackRemover(callback);
+    }
+
+    public static void reportLeak(LifecycleTracer tracer, Buffer recoveredBuffer) {
+        Set<Consumer<LeakInfo>> callbacks = CALLBACKS;
+        synchronized (callbacks) {
+            if (!callbacks.isEmpty()) {
+                LeakInfo info = new InternalLeakInfo(tracer, recoveredBuffer);
+                System.err.println("LEAK: "); // todo remove
+                info.forEach(trace -> { // todo remove
+                    trace.getTraceback().printStackTrace();
+                });
+                for (Consumer<LeakInfo> callback : callbacks) {
+                    callback.accept(info);
+                }
+            }
+        }
+    }
+
+    private static final class CallbackRemover implements AutoCloseable {
+        private final Consumer<LeakInfo> callback;
+
+        CallbackRemover(Consumer<LeakInfo> callback) {
+            this.callback = callback;
+        }
+
+        @Override
+        public void close() {
+            Set<Consumer<LeakInfo>> callbacks = CALLBACKS;
+            synchronized (callbacks) {
+                callbacks.remove(callback);
+            }
+            LEAK_DETECTION_ENABLED_UPDATER.getAndAddRelease(-1);
+        }
+    }
+
+    private static final class InternalLeakInfo implements LeakInfo {
+        private final LifecycleTracer tracer;
+        private final int bytesLeaked;
+        private Collection<TracePoint> cachedTrace;
+
+        InternalLeakInfo(LifecycleTracer tracer, Buffer recoveredBuffer) {
+            this.tracer = tracer;
+            bytesLeaked = recoveredBuffer.capacity();
+        }
+
+        @Override
+        public int bytesLeaked() {
+            return bytesLeaked;
+        }
+
+        @Override
+        public Iterator<TracePoint> iterator() {
+            return getTracePoints().iterator();
+        }
+
+        @Override
+        public Stream<TracePoint> stream() {
+            return getTracePoints().stream();
+        }
+
+        private Collection<TracePoint> getTracePoints() {
+            if (cachedTrace == null) {
+                cachedTrace = tracer.collectTraces();
```

Comment:
Do we need to ensure we wrap this so its umodifiable ? Otherwise `iterator()` might return an instance that can be mutated and so change the backing storage.

Reply:
Ah never mind... Seems like this is covered by the API:

https://github.com/netty/netty/pull/11725/files#diff-235cfac3e5bac48f1247b5316c80223d446deab922d0667f3f2488ac048cffd9R111